### PR TITLE
[ASCII-1531] Split tagger component into interface and implementation

### PR DESCRIPTION
**Pull request by GustavoCaso**
### What does this PR do?

Split the tagger component into interface package and implementation package 

### Motivation

As part of the Q2 OKR the ASC is trying to reduce the serverless binary overhead. As part of that effort we are removing certain components for serverless binary. 

We are creating noop implementation of those components. In the case of the tagger we first needed to split the component into interface and implementation. 

The great things about this is that not only the serverless binary benefits from this split as we can see in the import differences comment https://github.com/DataDog/datadog-agent/pull/24922#issuecomment-2069879151

For the record I computed some stats for the init duration of the serverless agent on this branch:

```
base:   min=478.33 ms, max=511.23 ms, avg=493.6 ms, median=490.38 ms
branch: min=460.34 ms, max=495.8 ms, avg=476.4 ms, median=477.56 ms
```

So splitting the tagger component removes around ~15ms of init time duration

### Additional Notes

Some of the changes I made are:
- Move types from the implementation package `comp/core/taggerimpl/...` into the interface package `comp/core/tagger`
- Update the Tagger interface. Unfortunately, that means I had to add the `Start` and `Stop` functions which is an anti-pattern for the component framework. I will try removing those in a future PR.
- Start using the tagger component interface rather than concrete struct types for some of the component interface functions.
- Update all references in the code. 

With this change we can see the serverless dependency graph no longer has the tagger dependency as part of dependencies:

Before it change:

```
goda tree -std "otlp=1(serverless=1(github.com/DataDog/datadog-agent/cmd/serverless:all))"  | grep tagger
                ├ github.com/DataDog/datadog-agent/comp/core/tagger/utils
            ├ github.com/DataDog/datadog-agent/comp/core/tagger
              ├ github.com/DataDog/datadog-agent/comp/core/tagger/api
              ├ github.com/DataDog/datadog-agent/comp/core/tagger/collectors
                ├ github.com/DataDog/datadog-agent/comp/core/tagger/utils ~
              ├ github.com/DataDog/datadog-agent/comp/core/tagger/local
                ├ github.com/DataDog/datadog-agent/comp/core/tagger/api ~
                ├ github.com/DataDog/datadog-agent/comp/core/tagger/collectors ~
                ├ github.com/DataDog/datadog-agent/comp/core/tagger/tagstore
                  ├ github.com/DataDog/datadog-agent/comp/core/tagger/api ~
                  ├ github.com/DataDog/datadog-agent/comp/core/tagger/collectors ~
                  ├ github.com/DataDog/datadog-agent/comp/core/tagger/subscriber
                    ├ github.com/DataDog/datadog-agent/comp/core/tagger/collectors ~
                    ├ github.com/DataDog/datadog-agent/comp/core/tagger/telemetry
                      ├ github.com/DataDog/datadog-agent/comp/core/tagger/collectors ~
                    ├ github.com/DataDog/datadog-agent/comp/core/tagger/types
                      ├ github.com/DataDog/datadog-agent/comp/core/tagger/collectors ~
                      └ github.com/DataDog/datadog-agent/comp/core/tagger/utils ~
                  ├ github.com/DataDog/datadog-agent/comp/core/tagger/telemetry ~
                  ├ github.com/DataDog/datadog-agent/comp/core/tagger/types ~
                ├ github.com/DataDog/datadog-agent/comp/core/tagger/telemetry ~
                ├ github.com/DataDog/datadog-agent/comp/core/tagger/types ~
              ├ github.com/DataDog/datadog-agent/comp/core/tagger/remote
                ├ github.com/DataDog/datadog-agent/comp/core/tagger/api ~
                ├ github.com/DataDog/datadog-agent/comp/core/tagger/collectors ~
                ├ github.com/DataDog/datadog-agent/comp/core/tagger/subscriber ~
                ├ github.com/DataDog/datadog-agent/comp/core/tagger/telemetry ~
                ├ github.com/DataDog/datadog-agent/comp/core/tagger/types ~
              ├ github.com/DataDog/datadog-agent/comp/core/tagger/replay
                ├ github.com/DataDog/datadog-agent/comp/core/tagger/api ~
                ├ github.com/DataDog/datadog-agent/comp/core/tagger/collectors ~
                ├ github.com/DataDog/datadog-agent/comp/core/tagger/tagstore ~
                ├ github.com/DataDog/datadog-agent/comp/core/tagger/telemetry ~
                ├ github.com/DataDog/datadog-agent/comp/core/tagger/types ~
                  ├ github.com/DataDog/datadog-agent/comp/core/tagger/collectors ~
                  ├ github.com/DataDog/datadog-agent/comp/core/tagger/types ~
              ├ github.com/DataDog/datadog-agent/comp/core/tagger/types ~
              ├ github.com/DataDog/datadog-agent/comp/core/tagger/utils ~
              ├ github.com/DataDog/datadog-agent/pkg/tagger/types
            ├ github.com/DataDog/datadog-agent/comp/core/tagger/collectors ~
                └ github.com/DataDog/datadog-agent/pkg/tagger/types ~
                ├ github.com/DataDog/datadog-agent/pkg/tagger/types ~
              ├ github.com/DataDog/datadog-agent/pkg/tagger/types ~
                    ├ github.com/DataDog/datadog-agent/comp/core/tagger ~
                ├ github.com/DataDog/datadog-agent/pkg/tagger/types ~
            ├ github.com/DataDog/datadog-agent/comp/core/tagger ~
            ├ github.com/DataDog/datadog-agent/comp/core/tagger/collectors ~
            ├ github.com/DataDog/datadog-agent/comp/core/tagger ~
            ├ github.com/DataDog/datadog-agent/comp/core/tagger/collectors ~
```

With this change:

```
➜ goda tree -std "otlp=1(serverless=1(github.com/DataDog/datadog-agent/cmd/serverless:all))"  | grep tagger
                ├ github.com/DataDog/datadog-agent/comp/core/tagger/utils
            ├ github.com/DataDog/datadog-agent/comp/core/tagger
              ├ github.com/DataDog/datadog-agent/comp/core/tagger/types
                ├ github.com/DataDog/datadog-agent/comp/core/tagger/utils ~
              ├ github.com/DataDog/datadog-agent/pkg/tagger/types
            ├ github.com/DataDog/datadog-agent/comp/core/tagger/types ~
                └ github.com/DataDog/datadog-agent/pkg/tagger/types ~
                ├ github.com/DataDog/datadog-agent/pkg/tagger/types ~
              ├ github.com/DataDog/datadog-agent/pkg/tagger/types ~
                    ├ github.com/DataDog/datadog-agent/comp/core/tagger ~
                      ├ github.com/DataDog/datadog-agent/comp/core/tagger/types ~
                ├ github.com/DataDog/datadog-agent/pkg/tagger/types ~
            ├ github.com/DataDog/datadog-agent/comp/core/tagger ~
            ├ github.com/DataDog/datadog-agent/comp/core/tagger/types ~
            ├ github.com/DataDog/datadog-agent/comp/core/tagger ~
            ├ github.com/DataDog/datadog-agent/comp/core/tagger/types ~
```

Both `tagger/utils` and `tagger` do not bring any additional dependencies 

### Describe how to test/QA your changes

There is no behavior change in the PR as it is mostly refactoring/moving code around

### CODE REVIEW

**Review by paulcacheux — APPROVED** [review]

**Review by ajgajg1134 — APPROVED** [review]
Looks good from agent-apm!

**Review by dmail — APPROVED** [review]
I discover go, datadog and basically everything related to the codebase so my review is very high level:
I understand what you want to do in this PR thanks to the description and changes looks coherent with that description 👌

**Review by songy23 — APPROVED** [review]
Looks good for OTel-owned files.

**Review by vickenty — APPROVED** [review]

**Review by Kaderinho — COMMENTED** [review]
One question regarding `Start`/`Stop` in the interface. The rest LGTM but I would rather wait for an approval from someone more familiar with this part of the code.

File: comp/core/tagger/component.go
Comment (Kaderinho) [reply comp/core/tagger/component.go:26]: I just saw your additional note 😓 then I think it's acceptable if we're gonna remove it later but you might want  thoughts from @hush-hush / @ogaca-dd

Comment (GustavoCaso) [reply comp/core/tagger/component.go:26]: Yeah, we try to move away from adding start and stop functions to the component interface. In this case I had to add them because it was part of the inner interface. I'm planning on removing on a separate PR

**Review by kangyili — APPROVED** [review]

**Review by jose-manuel-almaza — APPROVED** [review]

**Review by zhuminyi — APPROVED** [review]
LGTM

**Review by sblumenthal — APPROVED** [review]

**Review by zhuminyi — APPROVED** [review]

**Comment by GustavoCaso** [comment]
For the record I computed some stats for the init duration of the serverless agent on this branch:

base:   min=478.33 ms, max=511.23 ms, avg=493.6 ms, median=490.38 ms
branch: min=460.34 ms, max=495.8 ms, avg=476.4 ms, median=477.56 ms

So splitting the tagger component removes around ~15ms of init time duration

File: cmd/process-agent/command/main_common.go
Comment (louis-cqrl) [inline cmd/process-agent/command/main_common.go:34]: Do we still need this import since it is replace by `taggerimpl`

File: comp/core/tagger/component.go
Comment (louis-cqrl) [reply comp/core/tagger/component.go:26]: Is there a card created somewhere to avoid forgetting about removing them? If no could you create it ?

File: comp/core/tagger/global.go
Comment (louis-cqrl) [inline comp/core/tagger/global.go:20]: we had a pointer before, and now this is not a pointer is this intended ?

**Review by louis-cqrl — APPROVED** [review]
LGTM for files that ASC owns, left few questions

File: cmd/process-agent/command/main_common.go
Comment (GustavoCaso) [reply cmd/process-agent/command/main_common.go:34]: We reference the tagger package in multiple places:
https://github.com/DataDog/datadog-agent/blob/d4b130e60789a8e231e3228858727ee116bd0c7c/cmd/process-agent/command/main_common.go#L176

https://github.com/DataDog/datadog-agent/blob/d4b130e60789a8e231e3228858727ee116bd0c7c/cmd/process-agent/command/main_common.go#L203

File: comp/core/tagger/global.go
Comment (GustavoCaso) [reply comp/core/tagger/global.go:20]: Yes. We had to modify this to accept the tagger component interface so we can have a more unified experience when passing tagger types around.

File: cmd/process-agent/command/main_common.go
Comment (louis-cqrl) [reply cmd/process-agent/command/main_common.go:34]: okay didn't saw thx

File: comp/core/tagger/global.go
Comment (louis-cqrl) [reply comp/core/tagger/global.go:20]: okay perfect

**Comment by GustavoCaso** [comment]
/trigger-ci --variable RUN_ALL_BUILDS=true --variable RUN_KITCHEN_TESTS=true --variable RUN_E2E_TESTS=on --variable RUN_UNIT_TESTS=on --variable RUN_KMT_TESTS=on

File: comp/core/tagger/component.go
Comment (GustavoCaso) [reply comp/core/tagger/component.go:26]: I will create one

File: comp/core/tagger/taggerimpl/remote/tagger.go
Comment (GustavoCaso) [reply comp/core/tagger/taggerimpl/remote/tagger.go:256]: Great suggestion. I will add it to the PR

Comment (GustavoCaso) [reply comp/core/tagger/taggerimpl/remote/tagger.go]: Done https://github.com/DataDog/datadog-agent/pull/24922/commits/c69968d30a675c972950f5d9ada848b6a3bb4335

**Review by hush-hush — APPROVED** [review]

**Pull request merged**
